src/frontend/src/handler/privilege.rs

```diff
@@ -163,6 +129,100 @@ impl SessionImpl {
         Ok(())
     }
 
+    pub fn check_privileges_for_query(&self, query: &BoundQuery) -> Result<()> {
```

Comment:
We can avoid cloning the `query` if we still have a common method `check_privileges`, and then we can directly `resolve_query_privileges(query)` and then `check_privileges`.